taichi/program/context.h

```diff
@@ -109,7 +122,26 @@ struct RuntimeContext {
   void set_arg_ndarray(int arg_id,
                        intptr_t devalloc_ptr,
                        const std::vector<int> &shape) {
+    has_grad[arg_id] = false;
+    args[arg_id] = taichi_union_cast_with_different_sizes<uint64>(devalloc_ptr);
+    set_array_device_allocation_type(arg_id, DevAllocType::kNdarray);
+    TI_ASSERT(shape.size() <= taichi_max_num_indices);
+    size_t total_size = 1;
+    for (int i = 0; i < shape.size(); i++) {
+      extra_args[arg_id][i] = shape[i];
+      total_size *= shape[i];
+    }
+    set_array_runtime_size(arg_id, total_size);
+  }
+
+  void set_arg_ndarray(int arg_id,
```

Comment:
It seems the code of this two `set_arg_ndarray` overlap largely. Shall we merge them into one ?

Reply:
ok！

Reply:
It seems that the `has_grad[arg_id] = false` invoked by `ctx_->set_arg_ndarray(arg_id, ptr, arr.shape)`  will overwrite the `has_grad[arg_id] = true` before hand ?

Reply:
Is it possible that we invoke `ctx_->set_arg_ndarray(arg_id, ptr, arr.shape)` after `set_arg_ndarray_with_grad(actual_argument_slot, v_primal, v_grad)` for a same variable?

Reply:
I think it may overwrite if we do so.